icu4c/source/common/dictbe.cpp

```diff
@@ -1341,6 +1353,20 @@ CjkBreakEngine::divideUpDictionaryRange( UText *inText,
     if ((uint32_t)bestSnlp.elementAti(numCodePts) == kuint32max) {
         t_boundary.addElement(numCodePts, status);
         numBreaks++;
+    } else if (isPhraseBreaking) {
+        t_boundary.addElement(numCodePts, status);
+        numBreaks++;
+        int32_t prevIdx = numCodePts;
+
+        for (int32_t i = prev.elementAti(numCodePts); i > 0; i = prev.elementAti(i)) {
+            int length = prevIdx - i;
+            prevIdx = i;
+            // Skip the breakpoint if it belongs to the particle or Hiragana.
+            if (!fSkipSet.containsKey(inString.tempSubString(i, length))) {
```

Comment:
The naming of `numCodePts` suggests that this code counts code points, and that the length is a number of code points, but UnicodeString (and most of ICU) counts UTF-16 *code units*.

Please check (review the code and add a test case) that this works right, and/or add a comment about code unit vs. code point counting.

For example, here are the Hiragana/Katakana supplementary code points (each of these uses two UTF-16 code units):
[`[[:Hiragana:][:Katakana:]-[\u0000-\uFFFF]]`](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%5B%3AHiragana%3A%5D%5B%3AKatakana%3A%5D-%5B%5Cu0000-%5CuFFFF%5D%5D&g=&i=)

Reply:
Resolved. Change to use code unit